### ISSUE TITLE
Estimating KB memory usage

### ISSUE THREAD

**Issue by falcaopetri** [lajus/amie#34]
Hi,

`KB.load` tries to estimate the KB memory usage by calling `Runtime.getRuntime().freeMemory()` (Approach 1). In my experiments, reported memory varied across multiple runs (using the same KB). For small KBs, memory usage was negative.

It seems that `Runtime.getRuntime().freeMemory()` might not be always reliable (see https://stackoverflow.com/a/18366283). 

Also, I'm not sure if it is being correctly used in [KB.java](https://github.com/lajus/amie/blob/a2beeca2969202602d48f421ba682acc99a02ac2/kb/src/main/java/amie/data/KB.java#L871) since we are basically doing `curr_free_memory - prev_free_memory` to get non-free size:

```java

long memory = Runtime.getRuntime().freeMemory();
...
(Runtime.getRuntime().freeMemory() - memory) / 1000000
```
-------------------
An alternative suggested by the previous SO link is (Approach 2):

```java
long allocatedMemory = (Runtime.getRuntime().totalMemory()-Runtime.getRuntime().freeMemory());
```

Also, we might try calling `System.gc();` before each memory-size "snapshot" (https://stackoverflow.com/a/29431949) (Approach 1+GC, Approach 2+GC). 

Note that both Approach 2 and calling GC have their own issues.

---------------

### Experiment
Loading a 30MB KB (three times for each approach) gives me:

Approach 1 | Approach 2 | Approach 1+GC |Approach 2+GC
----|------|--------|------
120 | 459 | 121 | 359
131 | 433 | 301 | 359
124 | 457 | 97  | 359

Loading a 30kB KB (three times for each approach) gives me:

Approach 1 | Approach 2 | Approach 1+GC |Approach 2+GC
----|------|--------|-----|
-16 | 16 | -6 | 6
-16 | 16 | -7 | 6
-16 | 16 | -6  | 6

-------
I'm not sure if current memory reporting is actually bugged or worthy of being changed. Also, it seems that there is no fully reliable way to check a Java object size (at least not without adding dependencies and CLI arguments, https://stackoverflow.com/q/52353). Anyway, Approach 2+GC seems to be more stable. Should I submit a PR with the modification?

Regards,
Antonio.

**Comment by lgalarra** [comment]
Hi Antonio,

I had got negative numbers before, so I was aware of this bug but never corrected it because I never had the urgency of reporting accurate memory usage. That said, I think it is worth correcting or at least improving this estimate. From https://stackoverflow.com/questions/12807797/java-get-available-memory/18366283#18366283 I understand that this can happen because free memory is calculated from total memory where the latter can increase throughout time. Imagine the program had been assigned 1GB of RAM before loading the file and free memory is 20 MB. Immediately after, the JVM requests, let us say an additional GB, leaving us with 1044 MB of free memory. Let us say the loading reduces free memory to 512 MB.. the code will estimate 20MB - 512MB.

It seems that we can correct that estimating memory in line 832 as follows:

```
long presumableFreeMemory = Runtime.getRuntime().maxMemory() - allocatedMemory;
long allocatedMemory      = (Runtime.getRuntime().totalMemory()-Runtime.getRuntime().freeMemory());

```
And then do the same in line 868  

```
long presumableFreeMemoryAfter = Runtime.getRuntime().maxMemory() - allocatedMemory;
long allocatedMemoryAfter      = (Runtime.getRuntime().totalMemory()-Runtime.getRuntime().freeMemory());

long estimatedMemory = presumableFreeMemory - presumableFreeMemoryAfter;
```

As far as I understood Runtime.getRuntime().maxMemory() is constant, i.e., it is one of the bootstrapping parameters of the JVM. If the process was not assigned more memory between both sampling events, that expression is tantamount to what we have right now in the code.

Cheers,
Luis

**Comment by falcaopetri** [comment]
Hi,

I think your explanation about negative values seems correct.

> As far as I understood Runtime.getRuntime().maxMemory() is constant, i.e., it is one of the bootstrapping parameters of the JVM. If the process was not assigned more memory between both sampling events, that expression is tantamount to what we have right now in the code.

I agree that `maxMemory()` can be assumed to be constant. Even so, `estimatedMemory` is not equivalent to current implementation.

First, note that we do not actually need `presumableFreeMemoryAfter` or `presumableFreeMemory`,
since we can just do:
```java
long allocatedMemory = (Runtime.getRuntime().totalMemory()-Runtime.getRuntime().freeMemory());
long allocatedMemoryAfter = (Runtime.getRuntime().totalMemory()-Runtime.getRuntime().freeMemory());
long estimatedMemory = allocatedMemoryAfter - allocatedMemory;
```

Then, `estimatedMemory` becomes `(curr_totalMemory-curr_freeMemory) - (prev_totalMemory - prev_freeMemory)`.

Since `totalMemory` can change, it is different from current implementation `curr_freeMemory - prev_freeMemory`.

Regards,
Antonio.

**Comment by lgalarra** [comment]
Hi Antonio,

Yes, you are right. I have created a pull request: #39 

Once again, thanks for your contributions Antonio!

Best,
Luis

**Comment by falcaopetri** [comment]
PR merged in 651455d.
